File: Algo/Leetcode/073SetMatrixZeroes.cpp

//1. Easy Solution. Give O(mn) Time and O(m+n) Space.
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int m = matrix.size();
        if(m == 0) return;
        int n = matrix[0].size();
        
        vector<bool> row(m,true), col(n,true);
        for(int i=0;i<m*n; i++){
            int r = i/n;
            int c = i%n;
            if(matrix[r][c]==0){row[r]=false;col[c]=false;}
        }
        for(int i=0;i<m*n; i++){
            int r = i/n;
            int c = i%n;
            if(row[r]==false || col[c]==false){matrix[r][c]=0;}
        }
    }
};


//2. Use the first row & first column to store the values. O(1) Space.
//The assumption here is when there is a 0 in the smaller matrix, corresponding place in first row & col will be 0. So you can modify it.
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int row = matrix.size();
        int col = matrix[0].size();

        // consider first row and col
        bool row0 = false;
        bool col0 = false;
        for(int i=0; i<row; i++){
            if(matrix[i][0] == 0) col0 = true;
        }
        for(int j=0; j<col; j++){
            if(matrix[0][j] == 0) row0 = true;
        }
        // consider all other rows and col
        for(int i=1; i<row; i++){
            for(int j=1; j<col; j++){
                if(matrix[i][j] == 0){
                    matrix[0][j] = 0;
                    matrix[i][0] = 0;
                }
            }
        }
        // reset all other rows and cols
        for(int i=1; i<row; i++){
            if(matrix[i][0]==0){
                for(int j=1; j<col; j++) matrix[i][j] = 0;
            }
        }

        for(int j=1; j<col; j++){
            if(matrix[0][j] == 0){
                for(int i=1; i<row; i++) matrix[i][j] = 0;
            }
        }
        // reset first row and col
        if(row0){
            for(int j=0; j<col; j++) matrix[0][j] = 0;
        }

        if(col0){
            for(int i=0; i<row; i++) matrix[i][0] = 0;
        }

        return;
    }
};

File: Algo/Leetcode/073SetMatrixZeroes.py
class Solution(object):
    def setZeroes(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        m = len(matrix[0])
        row = [0] * n
        col = [0] * m

        for i in range(n):
            for j in range(m):
                if matrix[i][j] == 0:
                    row[i] = 1
                    col[j] = 1

        for i in range(n):
            for j in range(m):
                if row[i] == 1 or col[j] == 1:
                    matrix[i][j] = 0

        return matrix


# O(1) space
class Solution(object):
    def setZeroes(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        m = len(matrix[0])

        row0 = 1
        col0 = 1
        for j in range(m):
            if matrix[0][j] == 0:
                row0 = 0
        for i in range(n):
            if matrix[i][0] == 0:
                col0 = 0

        for i in range(1, n):
            for j in range(1, m):
                if matrix[i][j] == 0:
                    matrix[i][0] = 0
                    matrix[0][j] = 0

        for i in range(1, n):
            for j in range(1, m):
                if matrix[i][0] == 0 or matrix[0][j] == 0:
                    matrix[i][j] = 0

        for j in range(m):
            if row0 == 0:
                matrix[0][j] = 0
        for i in range(n):
            if col0 == 0:
                matrix[i][0] = 0

        return matrix